flink-python/pyflink/table/functions.py

```diff
@@ -138,6 +222,528 @@ def merge(self, accumulator, accumulators):
+                else:
+                    # this data has not been accumulated
+                    return
+                # get and remove current order in value_to_order_map
+                order = order_list.pop(0)
+                if len(order_list) == 0:
+                    del value_to_order_map[value]
+                else:
+                    value_to_order_map[value] = order_list
+            else:
+                order = args[1]
+
+            if order in order_to_value_map:
+                value_list = order_to_value_map[order]
+            else:
+                return
+
+            if value in value_list:
+                value_list.remove(value)
+                if len(value_list) == 0:
+                    del order_to_value_map[order]
+                else:
+                    order_to_value_map[order] = value_list
+
+            if value == prev_value:
+                start_key = prev_order
+                next_key = MIN_LONG_VALUE
+                for key in order_to_value_map:
+                    if start_key >= key > next_key:
+                        next_key = key
+
+                if next_key != MIN_LONG_VALUE:
+                    values = order_to_value_map[next_key]
+                    accumulator[0] = values[len(values) - 1]
+                    accumulator[1] = next_key
+                else:
+                    accumulator[0] = None
+                    accumulator[1] = None
+
+    def merge(self, accumulator, accumulators):
+        raise NotImplementedError("This function does not support merge.")
+
+
+class ListAggFunction(AggregateFunction):
+
+    def get_value(self, accumulator):
+        if accumulator[1] is None:
+            return None
+        else:
+            return accumulator[0].join(accumulator[1])
+
+    def create_accumulator(self):
+        # delimiter, values
+        return [',', None]
+
+    def accumulate(self, accumulator, *args):
+        if args[0] is not None:
+            if len(args) > 1:
+                accumulator[0] = args[1]
+            if accumulator[1] is None:
+                accumulator[1] = [args[0]]
+            else:
+                accumulator[1].append(args[0])
+
+    def retract(self, accumulator, *args):
+        raise NotImplementedError("This function does not support retraction.")
+
+
+class ListAggWithRetractAggFunction(AggregateFunction):
+
+    def get_value(self, accumulator):
+        values = [i for i in accumulator[0]]
+        if len(values) > 0:
+            return ','.join(values)
+        else:
+            return None
+
+    def create_accumulator(self):
+        # [list, retract_list]
+        return [ListView(), ListView()]
+
+    def accumulate(self, accumulator, *args):
+        if args[0] is not None:
+            accumulator[0].add(args[0])
+
+    def retract(self, accumulator, *args):
+        if args[0] is not None:
+            values = [i for i in accumulator[0]]
+            if args[0] in values:
```

Comment:
What about change the code to
`
try:
      values.remove(args(0))
      accumulator[0].clear()
      accumulator[0].add_all(values)
except ValueError:
      accumulator[1].add(args[0])
`